### ISSUE TITLE
False negative with kwargs and positional parameter

### ISSUE THREAD

**Issue by iafisher** [microsoft/pyright#10933]
```python
import subprocess
from typing import Any, Tuple

def sh2(cmd: str, check: bool = True, **kwargs: Any) -> Tuple[str, str]:
    proc = _sh(check=check, capture_output=True, **kwargs)
    return proc.stdout, proc.stderr

def _sh(cmd: str, **kwargs: Any) -> subprocess.CompletedProcess[str]:
    return subprocess.run(["/usr/bin/env", "bash", "-c", cmd], text=True, **kwargs)
```

`sh2` invokes `_sh` without passing the `cmd` positional argument, but pyright does not complain. I think it believes that `kwargs` could supply the `cmd` argument, but it can't because of the explicit `cmd` parameter in the signature of `sh2`. I do not believe it is possible to invoke `sh2` without triggering a `TypeError` at runtime.

Running pyright 1.1.405 on the command-line, on macOS.

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

The `kwargs` symbol has the type `dict[str, Any]`, so it could contain a key `cmd`. I see your point that in this particular code flow it would be possible for a type checker to determine that if `kwargs` cannot contain such a key at the entry of the `sh2` function and is not modified in any way between the entry and the point where it's used. That would require some pretty sophisticated analysis that goes beyond normal static type checking rules. None of the Python type checkers do this currently, to my knowledge.

I'm going to change this from a bug report to an enhancement request. I'm also going to reject the enhancement request at this time because I think this is an edge case that's unlikely to come up very often, and the time and effort it would take to add this feature is probably not justified. I'm willing to revisit this decision in the future if there's sufficient user signal to the contrary.
